### PR TITLE
entgql: move generate WhereInput logic to single hook

### PR DESCRIPTION
**Pull request by giautm**

### CODE REVIEW

File: entgql/extension.go [hunk 1]
```diff
@@ -196,12 +170,19 @@ func WithMapScalarFunc(scalarFunc func(*gen.Field, gen.Op) string) ExtensionOpti
...
+			Types:      map[string]*ast.Definition{},
+		},
+	}
 	for _, opt := range opts {
 		if err := opt(ex); err != nil {
 			return nil, err
 		}
 	}
+	ex.hooks = append(ex.hooks, ex.genSchemaHook())
 	ex.hooks = append(ex.hooks, removeOldAssets)
```
Comment (a8m) [inline entgql/extension.go:186]:
```suggestion
	ex.hooks = append(ex.hooks, ex.genSchemaHook(), removeOldAssets)
```

File: entgql/extension.go [hunk 2]
```diff
@@ -266,7 +249,18 @@ func (e *Extension) hasMapping(f *gen.Field) (string, bool) {
...
+			if strings.HasSuffix(m, ident) {
+				gqlNames = append(gqlNames, t)
+			}
+		}
+	}
+	if count := len(gqlNames); count == 1 {
+		return gqlNames[0], true
+	} else if count > 1 {
+		// NOTE(giautm): If there is more than 1 mapping,
+		// we will accept the one with the "Input" suffix by default.
```
Comment (a8m) [inline entgql/extension.go:261]:
```suggestion
		// If there is more than 1 mapping, we accept the one with the "Input" suffix.
```

File: entgql/extension.go [hunk 3]
```diff
@@ -283,63 +277,55 @@ func (e *Extension) hasMapping(f *gen.Field) (string, bool) {
...
+func (e *Extension) genSchemaHook() gen.Hook {
 	return func(next gen.Generator) gen.Generator {
 		return gen.GenerateFunc(func(g *gen.Graph) error {
 			if err := next.Generate(g); err != nil {
 				return err
 			}
 
+			if !(e.genSchema || e.genWhereInput) {
+				return nil
+			}
```
Comment (a8m) [inline entgql/extension.go:291]:
```suggestion
			if !e.genSchema && !e.genWhereInput {
				return nil
			}
```
Feels more intuitive to me.

File: entgql/internal/todoplugin/cmd/entgqlgen.go [hunk 4]
```diff
@@ -35,7 +36,14 @@ func main() {
...
 
 	exEntGQL, err := entgql.NewExtension(
+		entgql.WithConfigPath("./gqlgen.yml"),
+		entgql.WithWhereFilters(true),
 		entgql.WithSchemaGenerator(),
+		entgql.WithSchemaHook(func(g *gen.Graph, s *ast.Schema) error {
+			return nil
+		}),
+		// // Enable this option to write the output to a file
+		// entgql.WithSchemaPath("./entgql.graphql"),
```
Comment (a8m) [inline entgql/internal/todoplugin/cmd/entgqlgen.go:46]:
```suggestion
		// Enable this option to write the output to a file
		// entgql.WithSchemaPath("./entgql.graphql"),
```

**Review by a8m — COMMENTED** [review]
Thanks for the contribution @giautm. I added minor comments.

File: entgql/internal/todoplugin/cmd/entgqlgen.go [hunk 5]
```diff
@@ -35,7 +36,14 @@ func main() {
 	}
 
 	exEntGQL, err := entgql.NewExtension(
+		entgql.WithConfigPath("./gqlgen.yml"),
+		entgql.WithWhereFilters(true),
 		entgql.WithSchemaGenerator(),
+		entgql.WithSchemaHook(func(g *gen.Graph, s *ast.Schema) error {
+			return nil
+		}),
```
Comment (giautm) [inline entgql/internal/todoplugin/cmd/entgqlgen.go:44]: hello @crossworth, what do you think about this hook? Is it satisfied with this comment? https://github.com/ent/contrib/pull/267#issuecomment-1077900139

**Comment by giautm** [comment]
I can't see the lint comment. Don't know why, :-/
https://user-images.githubusercontent.com/12751435/161601023-c3cb59f4-7f85-489f-8464-3c54f6b25f8f.mov

File: entgql/internal/todoplugin/cmd/entgqlgen.go (hunk 5 above)
Comment (crossworth) [reply entgql/internal/todoplugin/cmd/entgqlgen.go:44]: Hello @giautm great work!
I have made a few small (and fast) tests and it worked without problems, I think this adds a lot of possibilities.

Maybe in the future we can collect the most common usages for schema hooks and add "helper functions".
Like adding/removing a field/directive, this should make even easier to the consumers.

I can't wait to use all this new features that you guys are adding.

**Comment by crossworth** [comment]
About the lint error, looks like it's about a panic:
` panic: load embedded ruleguard rules: rules/rules.go:13: can't load fmt`

I had to add the code below to `.golangci.yml` to make it work with GO 1.18.

```
run:
  go: '1.18'
```

**Comment by a8m** [comment]
> About the lint error, looks like it's about a panic: ` panic: load embedded ruleguard rules: rules/rules.go:13: can't load fmt`
> 
> I had to add the code below to `.golangci.yml` to make it work with GO 1.18.
> 
> ```
> run:
>   go: '1.18'
> ```

Thanks, @crossworth, please feel free to contribute and fix this. I failed back to go 1.17 for now to unblock our PRs.

**Pull request merged**